vpr/src/place/timing/PlacerCriticalities.h

```diff
@@ -0,0 +1,199 @@
+ *
+ * This process can be done incrementally, based on the modified connections/AtomPinIds
+ * returned by SetupTimingInfo. However, the set returned only reflects the connections
+ * changed by the last call to the timing info update.
+ *
+ * Therefore, if SetupTimingInfo is updated twice in succession without criticalities
+ * getting updated (update_enabled = false), the returned set cannot account for all
+ * the connections that have been modified. In this case, we flag `recompute_required`
+ * as false, and we recompute the criticalities for every connection to ensure that
+ * they are all up to date. Hence, each time update_setup_slacks_and_criticalities()
+ * is called, we assign `recompute_required` the opposite value of `update_enabled`.
+ *
+ * This class also maps/transforms the modified atom connections/pins returned by the
+ * timing info into modified clustered netlist connections/pins after calling
+ * update_criticalities(). The interface then enables users to iterate over this range
+ * via pins_with_modified_criticalities(). This is useful for incrementally re-calculating
+ * the timing costs.
+ *
+ * The criticalities of individual connections can then be queried by calling the
+ * criticality() member function.
+ *
+ * Implementation
+ * ==============
+ * To support incremental re-calculation, the class saves the last criticality exponent
+ * passed to PlacerCriticalities::update_criticalites(). If the next update uses the same
+ * exponent, criticalities can be incrementally updated. Otherwise, they must be re-calculated
+ * from scratch, since a change in exponent changes *all* criticalities.
+ *
+ * Calculating criticalities:
+ * All the raw setup slack values across a single clock domain are gathered
+ * and rated from the best to the worst in terms of criticalities. In order
+ * to calculate criticalities, all the slack values need to be non-negative.
+ * Hence, if the worst slack is negative, all the slack values are shifted
+ * by the value of the worst slack so that the value is at least 0. If the
+ * worst slack is positive, then no shift happens.
+ *
+ * The best (shifted) slack (the most positive one) will have a criticality of 0.
+ * The worst (shifted) slack value will have a criticality of 1.
+ *
+ * Criticalities are used to calculated timing costs for each connection.
+ * The formula is cost = delay * criticality.
+ *
+ * For a more detailed description on how criticalities are calculated, see
+ * calc_relaxed_criticality() in `timing_util.cpp`.
+ */
+class PlacerCriticalities {
+  public: //Types
+    typedef vtr::vec_id_set<ClusterPinId>::iterator pin_iterator;
+    typedef vtr::vec_id_set<ClusterNetId>::iterator net_iterator;
+
+    typedef vtr::Range<pin_iterator> pin_range;
+    typedef vtr::Range<net_iterator> net_range;
+
+  public: //Lifetime
+
+    ///@brief Allocates space for the timing_place_crit_ data structure.
+    PlacerCriticalities(const ClusteredNetlist& clb_nlist,
```

Comment:
Should explain the parameters here, particularly why the timing_info is a shared pointer (what is the usage/lifetime model)?